Core/TextureReplacer.cpp

```diff
@@ -553,123 +494,87 @@ enum class ReplacedImageType {
-	if (!fp) {
-		if (!ignoreError)
-			ERROR_LOG(G3D, "Error opening replacement texture file '%s'", level.file.c_str());
-		return false;
 	}
-
-	auto imageType = Identify(fp);
-	if (imageType == ReplacedImageType::ZIM) {
-		fseek(fp, 4, SEEK_SET);
-		good = fread(&level.w, 4, 1, fp) == 1;
-		good = good && fread(&level.h, 4, 1, fp) == 1;
-		int flags;
-		if (good && fread(&flags, 4, 1, fp) == 1) {
-			good = (flags & ZIM_FORMAT_MASK) == ZIM_RGBA8888;
+	// Turn the signature into a readable string that we can display in an error message.
+	*outMagic = std::string((const char *)magic, 4);
+	for (int i = 0; i < outMagic->size(); i++) {
+		if ((s8)(*outMagic)[i] < 32) {
+			(*outMagic)[i] = '_';
 		}
-	} else if (imageType == ReplacedImageType::PNG) {
-		png_image png = {};
-		png.version = PNG_IMAGE_VERSION;
-		if (png_image_begin_read_from_stdio(&png, fp)) {
-			// We pad files that have been hashrange'd so they are the same texture size.
-			level.w = png.width;
-			level.h = png.height;
-			good = true;
-		} else {
-			ERROR_LOG(G3D, "Could not load texture replacement info: %s - %s", level.file.ToVisualString().c_str(), png.message);
-		}
-		png_image_free(&png);
-	} else {
-		ERROR_LOG(G3D, "Could not load texture replacement info: %s - unsupported format", level.file.ToVisualString().c_str());
 	}
-	fclose(fp);
-
-	return good;
+	vfs->Rewind(openFile);
+	return IdentifyMagic(magic);
 }
 
-bool TextureReplacer::PopulateLevelFromZip(ReplacedTextureLevel &level, bool ignoreError) {
+bool TextureReplacer::PopulateLevel(ReplacedTextureLevel &level, bool ignoreError) {
 	bool good = false;
 
 	// Everything in here is within a lock, so we don't need to relock.
-	if (!level.zinfo || !level.zinfo->z || level.zi < 0) {
+	if (!level.fileRef) {
 		if (!ignoreError)
 			ERROR_LOG(G3D, "Error opening replacement texture file '%s' in textures.zip", level.file.c_str());
 		return false;
 	}
 
-	zip_file_t *zf = zip_fopen_index(level.zinfo->z, level.zi, 0);
-	if (!zf)
+	File::FileInfo info;
+	if (!vfs_->GetFileInfo(level.fileRef, &info)) {
 		return false;
+	}
+	size_t zsize = info.size;
 
-	auto imageType = Identify(zf);
-	zip_fclose(zf);
+	VFSOpenFile *file = vfs_->OpenFileForRead(level.fileRef);
+	if (!file) {
+		return false;
+	}
+
+	std::string magic;
+	auto imageType = Identify(vfs_, file, &magic);
 
-	zf = zip_fopen_index(level.zinfo->z, level.zi, 0);
 	if (imageType == ReplacedImageType::ZIM) {
 		uint32_t ignore = 0;
-		good = zip_fread(zf, &ignore, 4) == 4;
-		good = good && zip_fread(zf, &level.w, 4) == 4;
-		good = good && zip_fread(zf, &level.h, 4) == 4;
-		int flags;
-		if (good && zip_fread(zf, &flags, 4) == 4) {
-			good = (flags & ZIM_FORMAT_MASK) == ZIM_RGBA8888;
-		}
+		struct ZimHeader {
+			uint32_t magic;
+			uint32_t w;
+			uint32_t h;
+			uint32_t flags;
+		} header;
+		good = vfs_->Read(file, &header, sizeof(header)) == sizeof(header);
+		good = (header.flags & ZIM_FORMAT_MASK) == ZIM_RGBA8888;
```

Comment:
I think this is missing actually setting `level.w` and `level.h` to the loaded w/h values.  They're read but not used.  So this probably breaks zims.

-[Unknown]

Reply:
Oops, fixing.